### PR TITLE
Blog app - integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by shegy28**
### Add  integration specs for views and fix n+1 problems

**In this branch we implemented the following:**

- Made sure that the N+1 problem is solved when fetching all posts and their comments for a user.
-  Used Capybara to write some integration tests for the:
-  User Index page
-  User Show page
-  User Post Index page
-  Post Show page

Screenshots below:
n+1 issue 1

Screenshot (5)

### CODE REVIEW

File: spec/views/users_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      @user3 = User.create(
+        name: 'Tom',
+        photo: 'https://img.freepik.com/free-vector/mysterious-mafia-man-smoking-cigarette_52683-34828.jpg?w=1480&t=st=1674661161~exp=1674661761~hmac=e529680000ea966150e5ea3b38241a0d8c9582faf23dac8af61ce1785cf27838',
+        bio: 'Teacher from Mexico.',
+        posts_counter: 0
+      )
+    ]
+    visit('/')
+  end
+  describe 'users#index' do
```
Comment (rloterh) [inline spec/views/users_index_spec.rb:29]: - [ ] Kindly also include test for the following:
  * When I click on a user, I am redirected to that user's show page.

File: spec/views/users_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      ),
+      @user3 = User.create(
+        name: 'Tom',
+        photo: 'https://img.freepik.com/free-vector/mysterious-mafia-man-smoking-cigarette_52683-34828.jpg?w=1480&t=st=1674661161~exp=1674661761~hmac=e529680000ea966150e5ea3b38241a0d8c9582faf23dac8af61ce1785cf27838',
+        bio: 'Teacher from Mexico.',
+        posts_counter: 0
+      )
+    ]
+  end
+
```
Comment (rloterh) [inline spec/views/users_show_spec.rb:35]: - [ ] Similarly, please ensure provide the following test cases for the _User show_ view:
   * I can see the user's profile picture.
   * I can see the user's username.
   * I can see the number of posts the user has written.
   * I can see the user's bio.
   * I can see the user's first 3 posts.
   * I can see a button that lets me view all of a user's posts.
   * When I click a user's post, it redirects me to that post's show page.
   * When I click to see all posts, it redirects me to the user's post's index page.

File: spec/views/posts_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        author: @user1, title: 'List of all posts', text: 'This is my first post', comments_counter: 0, likes_counter: 0
+      ),
+      @post2 = Post.create(
+        author: @user1, title: 'Football', text: 'This is my second post', comments_counter: 0, likes_counter: 0
+      ),
+      @post3 = Post.create(
+        author: @user1, title: 'Current Affairs', text: 'This is my third post', comments_counter: 0, likes_counter: 0
+      )
+    ]
+
```
Comment (rloterh) [inline spec/views/posts_index_spec.rb:30]: - [ ] Again, let's ensure to provide the following test scenarios for the _post index view_:
   * I can see a section for pagination if there are more posts than fit on the view.
   * When I click on a post, it redirects me to that post's show page.

File: spec/views/posts_show_spec.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+      ),
+      @post2 = Post.create(
+        author: @user2,
+        title: 'Hello',
+        text: 'This is my second post',
+        comments_counter: 0,
+        likes_counter: 0
+      )
+    ]
+
```
Comment (rloterh) [inline spec/views/posts_show_spec.rb:34]: - [ ] Lastly, please provide the below test scenario for _post show_ view:
   * I can see the comment each commenter left.

File: spec/views/posts_show_spec.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    visit user_post_path(@users.first, @posts.first)
+  end
+
+  describe '#PostShowPage' do
+    it 'Should display post title' do
+      expect(page).to have_content(@post1.title)
```
Comment (rloterh) [inline spec/views/posts_show_spec.rb:40]: - [ ] One of the integration specs for _post show_ view is failing. Kindly ensure that all test cases are passing!

**Failing test case:**
image

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Chimezie, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project, but you are almost there! 
🏁 🚦    🏃 🌴  🏫 🌳 🏘️ 

## Highlights 👍  
- N + 1 problem addressed!

## Required Changes ♻️
_Kindly check the inline comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

**You can also consider:**
- Kindly include appropriate instructions in your ReadMe on how to run the _Rspec_ specs for this app, under the _Getting Started_ section. Implementing the above change(s) will make your ReadMe/ repository stand out even better! 😉


Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Reem-lab — APPROVED** [review]
# Hi Team,👋🏻👋🏻

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 🔥🔥💯💯


## Highlights 🔥🔥🔥🔥
- Followed Gitflow ✅
- Great working on the functionality 🔥🔥
- ALl the requested changes are applied well ✅🤩




## Optional suggestions



_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**